beacon-chain/sync/pending_attestations_queue.go

```diff
@@ -0,0 +1,160 @@
+
+			// Delete the missing block root key from pending attestation queue, a node will not request for the block again.
+			delete(s.blkRootToPendingAtts, bRoot)
+		} else {
+			// The pending attestation's missing block has not arrived yet.
+			log.WithField("blockRoot", hex.EncodeToString(bytesutil.Trunc(bRoot[:]))).Info("Requesting block for pending attestation")
+
+			// Start with a random peer to query, but choose the first peer in our unsorted list that claims to
+			// have a head slot newer than the pending attestation's target boundary slot.
+			pid := pids[rand.Int()%len(pids)]
+			targetSlot := helpers.SlotToEpoch(attestations[0].Aggregate.Data.Target.Epoch)
+			for _, p := range pids {
+				if cs, _ := s.p2p.Peers().ChainState(p); cs != nil && cs.HeadSlot >= targetSlot {
+					pid = p
+					break
+				}
+			}
+
+			req := [][32]byte{bRoot}
+			if err := s.sendRecentBeaconBlocksRequest(ctx, req, pid); err != nil {
+				traceutil.AnnotateError(span, err)
+				log.Errorf("Could not send recent block request: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// This defines how pending attestations is saved in the map. The key is the
+// root of the missing block. The value is the list of pending attestations
+// that voted for that block root.
+func (s *Service) savePendingAtt(att *ethpb.AggregateAttestationAndProof) {
+	s.pendingAttsLock.Lock()
+	defer s.pendingAttsLock.Unlock()
+
+	root := bytesutil.ToBytes32(att.Aggregate.Data.BeaconBlockRoot)
+
+	_, ok := s.blkRootToPendingAtts[root]
+	if !ok {
+		s.blkRootToPendingAtts[root] = []*ethpb.AggregateAttestationAndProof{att}
+	} else {
+		s.blkRootToPendingAtts[root] = append(s.blkRootToPendingAtts[root], att)
+	}
+}
+
+// This validates the pending attestations in the queue are still valid.
+// If not valid, a node will remove it in the queue in place. The validity
+// check specifies the pending attestation could not fall one epoch behind
+// of the current slot.
+func (s *Service) validatePendingAtts(ctx context.Context, slot uint64) {
+	s.pendingAttsLock.Lock()
+	defer s.pendingAttsLock.Unlock()
+
+	ctx, span := trace.StartSpan(ctx, "validatePendingAtts")
+	defer span.End()
+
+	for bRoot, atts := range s.blkRootToPendingAtts {
+		for i, att := range atts {
+			if slot >= att.Aggregate.Data.Slot+params.BeaconConfig().SlotsPerEpoch {
+				// Remove the pending attestation from the list in place.
+				if len(atts) > 1 {
+					atts[len(atts)-1], atts[i] = atts[i], atts[len(atts)-1]
+				}
+				atts = atts[:len(atts)-1]
+				numberOfAttsNotRecovered.Inc()
+			}
+		}
+		s.blkRootToPendingAtts[bRoot] = atts
+
+		// If the pending attestations list of a given block root is empty,
+		// a node will remove the key from the map to avoid dangling keys.
+		if len(s.blkRootToPendingAtts[bRoot]) == 0 {
+			delete(s.blkRootToPendingAtts, bRoot)
+			numberOfBlocksNotRecoveredFromAtt.Inc()
+		}
+	}
+}
+
+// This returns the current slot based on the time now.
+func (s *Service) currentSlot() uint64 {
```

Comment:
Can you reuse this? https://github.com/prysmaticlabs/prysm/blob/c2fbb40909d7bb2a2af493aa8fd1f125c1d0bfde/beacon-chain/blockchain/process_block_helpers.go#L265

Reply:
This depends on the service struct: `s.genesisTime` 

Do you think it's a better pattern if I export `CurrentSlot()` from block chain service and use it in other services?

Reply:
Yeah, you are using s.genesisTime already with the `s.chain.GenesisTime()`

https://github.com/prysmaticlabs/prysm/blob/0d58a463a1709ff9f074573453e99e3e9a89fa23/beacon-chain/blockchain/chain_info.go#L181-L183

Reply:
done